tapgarden/custodian_test.go

```diff
@@ -266,6 +300,80 @@ func TestCustodianNewAddr(t *testing.T) {
 	})
 }
 
+// TestBookAssetSyncer makes sure that addresses can be created for assets
+// not yet known to the address book.
+func TestBookAssetSyncer(t *testing.T) {
+	t.Parallel()
+
+	h := newHarness(t, nil)
+	require.NoError(t, h.c.Start())
+	t.Cleanup(func() {
+		require.NoError(t, h.c.Stop())
+	})
+	h.assertStartup()
+
+	ctx := context.Background()
+	proofCourierAddr := address.RandProofCourierAddr(t)
+
+	// Start a background goroutine to add assets that have been
+	// fetched from the asset syncer to the address book, to mimic a
+	// universe sync.
+	quitAssetWatcher := make(chan struct{})
+	insertAssetInfo(
+		t, ctx, quitAssetWatcher, h.tapdbBook, h.syncer,
+	)
+
+	// Address creation should fail for unknown assets.
+	newAsset := asset.RandAsset(t, asset.Type(test.RandInt31n(2)))
+	_, err := h.addrBook.NewAddress(
+		ctx, newAsset.ID(), 1, nil, proofCourierAddr,
+	)
+	require.ErrorContains(t, err, "unknown asset")
+
+	// If we add the asset to the asset syncer, address creation should
+	// succeed.
+	h.syncer.AddAsset(*newAsset)
+	go func() {
+		<-h.keyRing.ReqKeys
+		<-h.keyRing.ReqKeys
+	}()
+	newAddr, err := h.addrBook.NewAddress(
+		ctx, newAsset.ID(), 1, nil, proofCourierAddr,
+	)
+	require.NoError(t, err)
+	require.NotNil(t, newAddr)
+
+	numAddrs := 1
+	validAddrs := fn.MakeSlice(newAddr)
+
+	// TODO(jhb): remove asset from syncer, assert that address creation
```

Comment:
what is the expected effect of removing an asset from the syncer?

Reply:
Here, the `address.Book` should still succeed to make an address, as `QueryAssetGroup` should find the asset genesis (since the address should have been imported after the `AddAsset` call above.